hist/histv7/inc/ROOT/RHistImpl.hxx

```diff
@@ -227,122 +257,467 @@ namespace Internal {
-         range[1][I] = std::get<I>(axes).end_with_overflow();
-      else
-         range[1][I] = std::get<I>(axes).end();
-      RFillIterRange<I - 1, AXES>()(range, axes, over);
+      constexpr const int thisAxis = NDIMS - I - 1;
+      if ((!std::get<thisAxis>(axes).CanGrow()) && (virtualBins[thisAxis] == 0)) {
+         localBins[thisAxis] = RAxisBase::kUnderflowBin;
+      } else if ((!std::get<thisAxis>(axes).CanGrow()) && (virtualBins[thisAxis] == (std::get<thisAxis>(axes).GetNBins() - 1))) {
+         localBins[thisAxis] = RAxisBase::kOverflowBin;
+      } else {
+         const int regular_bin_offset = -static_cast<int>(std::get<thisAxis>(axes).CanGrow());
+         localBins[thisAxis] = virtualBins[thisAxis] - regular_bin_offset;
+      }
+      RVirtualBinsToLocalBins<I - 1, NDIMS, BINS, AXES>()(localBins, axes, virtualBins);
    }
 };
 
-enum class EBinCoord {
-   kBinFrom,   ///< Get the lower bin edge
-   kBinCenter, ///< Get the bin center
-   kBinTo      ///< Get the bin high edge
+/// Recursively converts local axis bins from the standard `kUnderflowBin`/`kOverflowBin` for under/overflow
+/// bin indexing convention, to a "virtual bin" convention where the underflow bin
+/// has index `0` and the overflow bin has index `N+1` where `N` is the axis'
+/// number of regular bins.
+///
+/// For growable axes, subtract 1 from regular indices so that the indexing
+/// convention remains zero-based (this means that there will be no "holes" in
+/// global binning, which matters more than the choice of regular index base)
+template <int I, int NDIMS, typename BINS, class AXES>
+struct RLocalBinsToVirtualBins;
+
+template <int NDIMS, typename BINS, class AXES>
+struct RLocalBinsToVirtualBins<-1, NDIMS, BINS, AXES> {
+   void operator()(BINS & /*virtualBins*/, const AXES & /*axes*/, const BINS & /*localBins*/) const
+   {}
 };
 
-template <int I, int NDIM, class COORD, class AXES>
-struct RFillBinCoord;
+template <int I, int NDIMS, typename BINS, class AXES>
+struct RLocalBinsToVirtualBins {
+   void operator()(BINS &virtualBins, const AXES &axes, const BINS &localBins) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      switch (localBins[thisAxis]) {
+         case RAxisBase::kUnderflowBin:
+               virtualBins[thisAxis] = 0; break;
+         case RAxisBase::kOverflowBin:
+               virtualBins[thisAxis] = std::get<thisAxis>(axes).GetNBins() - 1; break;
+         default:
+               virtualBins[thisAxis] = localBins[thisAxis] - static_cast<int>(std::get<thisAxis>(axes).CanGrow());
+      }
+      RLocalBinsToVirtualBins<I - 1, NDIMS, BINS, AXES>()(virtualBins, axes, localBins);
+   }
+};
+
+/// Recursively converts bin coordinates to the corresponding local axis bins
+/// from the standard `kUnderflowBin`/`kOverflowBin` for under/overflow bin indexing convention.
+template <int I, int NDIMS, typename BINS, typename COORD, class AXES>
+struct RCoordsToLocalBins;
```

Comment:
Since this is just calling `FindBin` on every axis and producing local coordinates, maybe `RFindLocalBins` would be a clearer name?